File: bin2c.js
const { readFileSync, openSync, writeSync, closeSync } = require('fs');
const { join } = require('path');

const [input, output, name] = process.argv.slice(2);
if (!input || !output || !name) {
    console.log('Usage: node bin2c.js [input] [output] [name]');
    process.exit(-1);
}

const fd = openSync(output, 'w');

writeSync(fd, `#include <stdint.h>\n`);
writeSync(fd, `\n`);
writeSync(fd, `uint8_t ${name}[] = {\n`);

let buf = readFileSync(input, { encoding: null });
while (buf.length > 0) {
    const line = Array.prototype.slice.apply(buf.slice(0, 16))
        .map(b => `0x${b.toString(16).padStart(2, '0')}`)
        .join(', ');

    writeSync(fd, `\t${line},\n`);

    buf = buf.slice(16);
}

writeSync(fd, `};\n`);
writeSync(fd, `\n`);
writeSync(fd, `uint32_t ${name}_len = sizeof(${name});\n`);
closeSync(fd);

File: bin2c.py
import sys


def main():
    if len(sys.argv) != 4:
        print("Usage: python bin2c.py [input] [output] [name]")
        sys.exit(-1)

    input_path, output_path, name = sys.argv[1], sys.argv[2], sys.argv[3]

    with open(output_path, "w", newline="\n") as out:
        out.write("#include <stdint.h>\n\n")
        out.write(f"uint8_t {name}[] = {{\n")

        with open(input_path, "rb") as f:
            while True:
                chunk = f.read(16)
                if not chunk:
                    break
                line = ", ".join(f"0x{b:02x}" for b in chunk)
                out.write(f"\t{line},\n")

        out.write("};\n\n")
        out.write(f"uint32_t {name}_len = sizeof({name});\n")


if __name__ == "__main__":
    main()
